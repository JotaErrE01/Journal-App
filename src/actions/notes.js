import { addDoc, collection, updateDoc, doc, deleteDoc } from "firebase/firestore";
import Swal from "sweetalert2";
import { db } from "../firebase/config";
import { loadNotes } from "../helpers/LoadNotes";
import { fileUpload } from "../hooks/fileUpload";
import types from "../types/types";

const startNewNote = () => {
    return async ( dispatch, getState ) => {

        const { uid } = getState().auth;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const docRef = await addDoc(collection(db, `${ uid }/journal/notes`), newNote);
            // await Swal.fire( 'Nota Agregada', 'La nota fue agregada con exito', 'success' );

            // console.log(docRef);

            dispatch( activeNote( docRef.id, newNote ) );

            dispatch( addNewNote({ id: docRef.id, ...newNote }) );

        } catch (e) {
            console.error("Error adding document: ", e);
        }
    }
};

const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

const addNewNote = note => ({
    type: types.notesAddNew,
    payload: note
})

const startLoadinNotes = () => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        const notes = await loadNotes( uid );
        dispatch( setNotes( notes ) );
    }
};

const setNotes = notes => ({
    type: types.notesLoad,
    payload: notes
});

const startSaveNote = note => {
    return async (dispatch, getState) => {

        Swal.fire({ 
            title: 'Uploading...', 
            text: 'Please Wait',
            allowOutsideClick: false, 
            didOpen: () => {
                Swal.showLoading();
            }
        })

        const { uid } = getState().auth;

        const noteFiresotore = { ...note };

        if( !note.url ) delete noteFiresotore.url;
        
        // Agregar Image
        if( note.file ){
            const fileUrl = await fileUpload( note.file ); // sube el archivo a cloud dinary

            noteFiresotore.url = fileUrl;

            // eliminar datos no soportados por firebase
            delete noteFiresotore.file;
        }

        // eliminar el id del objeto
        delete noteFiresotore.id;

        try {
            const docRef = doc(db, `${ uid }/journal/notes/${ note.id }`); // referencia al doc

            // actualizamos el doc
            await updateDoc(docRef, noteFiresotore);

            dispatch( refreshNote( { ...noteFiresotore, id: note.id } ) );

            Swal.fire("Nota Actualizada", "Nota actualizada correctamente", "success");

        } catch (error) {
            console.log(error);
            Swal.fire("Error", "Error al actualizar la nota", "error");
        }
        
    }    
}

const refreshNote = note => ({
    type: types.notesUpdated,
    payload: note
})

// const startUploading = ( file ) => {
//     return async (dispactch, getState) => {
//         const { active: note } = getState().notes;
        
//         const url = URL.createObjectURL(file);

//         console.log(url);

//         console.log(URL.revokeObjectURL( url ));
        
//         dispactch( activeNote( note.id, { ...note, url } ) );
//     }
// }

const startDeleting = id => {
    return async (dispactch, getState) => {

        const uid = getState().auth.uid;
          
        const result = await Swal.fire({
            title: 'Estas seguro?',
            text: "Los datos eliminados no se podran recuperar!",
            icon: 'warning',
            showCancelButton: true,
            confirmButtonText: 'SÃ­, Eliminar!',
            cancelButtonText: 'No, cancelar!',
            reverseButtons: true
        })

        if (result.isConfirmed) {

            // eliminar de firesatore
            await deleteDoc(doc(db, `${ uid }/journal/notes/${ id }`));

            // eliminar nota de la memoria
            dispactch( deleteNote( id ) );

            Swal.fire(
                'Eliminado!',
                'Tu nota ha sido eliminada exitosamente.',
                'success'
            )

        } else if (result.dismiss === Swal.DismissReason.cancel) {
            Swal.fire(
                'Cancelado',
                'Nota no eliminada',
                'error'
            )
        }

    }
}

const deleteNote = id => ({
    type: types.notesDelete,
    payload: id
})

const noteLogout = () => ({
    type: types.notesLogoutCleaning
})

export {
    startNewNote,
    activeNote,
    setNotes,
    startLoadinNotes,
    startSaveNote,
    // startUploading,
    startDeleting,
    noteLogout
}