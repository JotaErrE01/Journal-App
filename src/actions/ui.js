import types from "../types/types";


const setError = msg => ({
    type: types.uiSetError,
    payload: msg
})

const removeError = () => ({
    type: types.uiRemoveError
})

const uiStartLoading = () => ({
    type: types.uiStartLoading
})

const uiFinishLoading = () => ({
    type: types.uiFinishLoading
})

export {
    setError,
    removeError,
    uiStartLoading,
    uiFinishLoading
}