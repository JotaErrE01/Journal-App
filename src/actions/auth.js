import { signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from "firebase/auth";
import { auth, googleAuthProvider } from '../firebase/config';
import types from "../types/types";
import { uiFinishLoading, uiStartLoading } from "./ui";
import Swal from 'sweetalert2';
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    return async dispatch => {
        dispatch( uiStartLoading() );
        try {
            const { user } = await signInWithEmailAndPassword( auth, email, password );
            dispatch( login( user.uid,  user.displayName ) );
            dispatch( uiFinishLoading() );
        } catch (error) {
            dispatch( uiFinishLoading() );
            Swal.fire('Error', error.message, 'error')
        }
    }
}

export const starRegisterWithEmailPassworName = ( email, password, name ) => {
    return async dispatch => {
        try {
            
            const userCredentials = await createUserWithEmailAndPassword( auth, email, password ); // esta funcin aparte de crear tambien logea al usuario en firebase

            const user =  userCredentials.user;

            await updateProfile( user, {
                displayName: name
            } );

            dispatch( login( user.uid, user.displayName ) );

        } catch (error) {
            Swal.fire('Error', error.message, 'error')
        }

    }
}

export const startGoogleLogin = () => {

    return dispatch => {
        signInWithPopup(auth, googleAuthProvider)
            .then( userCredential => {
                dispatch( login, userCredential.user.uid, userCredential.user.displayName );
            } )
            .catch( error => {
                console.log(error);
            } );
    }
}

export const login = ( uid, name ) => {
    return {
        type: types.login,
        payload: {
            uid,
            name
        }
    }
}

export const startLogout = () => {
    return async dispatch => {
        try {
            await signOut( auth );
            dispatch( logout() );
            dispatch( noteLogout() );
        } catch (error) {
            console.log(error);
        }
    }
}

export const logout = () => ({
    type: types.logout
})