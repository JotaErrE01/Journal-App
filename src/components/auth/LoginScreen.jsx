import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';


const LoginScreen = () => {

    const dispatch = useDispatch();
    const error = useSelector(state => state.ui.msgError);
    const loading = useSelector(state => state.ui.loading);

    const [ { email, password }, handleInputChange ] = useForm({
        email: '',
        password:''
    });

    const handleLogin = e => {
        e.preventDefault();

        // validar si hay campos  vacios
        if(email.trim() === '' || password.trim() === '' ){
            dispatch( setError( 'Todos los campos son obligatorios' ) );
            return;
        }

        if(!validator.isEmail(email)){
            dispatch( setError( 'El email no es valido' ) );
            return;
        }

        dispatch( removeError() );
        dispatch( startLoginEmailPassword( email, password ) );
    }

    const handleGoogleLogin = () => {
        dispatch( removeError() );
        dispatch( startGoogleLogin() );
    }

    return (
        <div className="animate__animated animate__fadeIn animate__faster">
            <h3 className='auth__title mb-5'>Login</h3>

            <form
                onSubmit={handleLogin}
            >

                {
                    error &&
                        <div className="auth__alert-error">
                            { error }
                        </div>
                }

                <input  
                    type="text" 
                    placeholder="Email"
                    name="email"
                    value={email}
                    className='auth__input mb-5'
                    autoComplete="off"
                    onChange={handleInputChange}
                />

                <input 
                    type="password" 
                    placeholder="Password"
                    name="password"
                    value={password}
                    className='auth__input mb-5'
                    autoComplete="off"
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className='btn btn-primary pointer btn-block'
                    disabled={ loading }
                >Login</button>

                <hr />

                <div className='auth__social-networks'>
                    <p>Login With Social Networks</p>

                    <div 
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to='/auth/register' className="link" >Create New Account</Link>
            </form>
        </div>
    )
}

export default LoginScreen;