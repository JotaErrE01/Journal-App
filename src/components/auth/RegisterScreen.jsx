import React from 'react'
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { useSelector } from 'react-redux';
import { starRegisterWithEmailPassworName } from '../../actions/auth';

const RegisterScreen = () => {

    const dispacth = useDispatch();
    const error = useSelector(state => state.ui.msgError);

    const [ usuario, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });

    const { name, email, password, password2 } = usuario;

    const handleRegister = e => {
        e.preventDefault();

        if(isFormValid()){
            dispacth( removeError() );
            dispacth( starRegisterWithEmailPassworName( email, password, name ) );
        }
    }

    // validar si el formulario es valido
    const isFormValid = () => {

        if( name.trim() === '' ) {
            dispacth( setError( 'Name is required' ) );
            return false;
        }

        if( !validator.isEmail(email) ){
            dispacth( setError( 'Email is not valid' ) );
            return false;
        }

        if( password.length < 5 || password !== password2 ){
            dispacth( setError( 'Password should be at least 5 characters and match each others' ) );
            return false;
        }

        return true;
    }

    return (
        <div
            className="animate__animated animate__fadeIn animate__faster"
        >
            <h3 className='auth__title mb-5'>Register</h3>

            <form
                onSubmit={handleRegister}
            >

                {
                    error &&
                        <div className="auth__alert-error">
                            { error }
                        </div>
                }

                <input  
                    type="text" 
                    placeholder="Name"
                    name="name" 
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={name}
                    onChange={handleInputChange}
                />

                <input  
                    type="text" 
                    placeholder="Email"
                    name="email" 
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input 
                    type="password" 
                    placeholder="Password"
                    name="password"
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={password}
                    onChange={handleInputChange}
                />

                <input 
                    type="password" 
                    placeholder="Conifrm Password"
                    name="password2"
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={password2}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className='btn btn-primary pointer btn-block mb-5'
                >Register</button>

                <Link to='/auth/login' className="link" >Already Register</Link>
            </form>
        </div>
    )
}

export default RegisterScreen;