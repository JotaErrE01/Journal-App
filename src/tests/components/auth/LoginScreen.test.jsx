import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import LoginScreen from "../../../components/auth/LoginScreen";
import '@testing-library/jest-dom';
import { startGoogleLogin, startLoginEmailPassword } from "../../../actions/auth";

jest.mock( "../../../actions/auth", () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}) );

// const middleware = [ thunk ];
const mockStore = configureStore([ thunk ]);
const initState = {
    auth: { },
    ui: {
        loading: false,
        msg: null
    }
}

let store;
let wrapper;

describe('Pruebas en <LoginScreen />', () => {

    beforeEach(() => {
        store = mockStore( initState );
        store.dispatch = jest.fn();
        wrapper = render(

            <Provider store={ store } >
                <MemoryRouter>
                    <LoginScreen />
                </MemoryRouter>
            </Provider>
            
        );
    })


    test('Debe de hacer match con el snapshot', () => {
        
        expect( wrapper.container ).toMatchSnapshot();

    })

    test('Debe de disparar la accion de startGoogleLogin y de removeError', () => {

        userEvent.click( screen.getByText( 'Sign in with google' ) );

        expect( startGoogleLogin ).toHaveBeenCalled();

    })

    test('Debe de disparar el startLogin con los respectivos argumentos', () => {
        
        const { container } = wrapper;
        const [ email, pass ] = container.querySelectorAll( '.auth__input' );

        userEvent.type( email, 'correo@correo.com' );
        userEvent.type( pass, '123456' );
        userEvent.click( container.querySelector( '.btn-block' ) );

        expect( startLoginEmailPassword ).toHaveBeenCalled();
        expect( startLoginEmailPassword ).toHaveBeenCalledWith('correo@correo.com', '123456');

    })

})