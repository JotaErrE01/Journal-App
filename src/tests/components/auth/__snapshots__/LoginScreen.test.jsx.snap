// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="animate__animated animate__fadeIn animate__faster"
  >
    <h3
      class="auth__title mb-5"
    >
      Login
    </h3>
    <form>
      <input
        autocomplete="off"
        class="auth__input mb-5"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="auth__input mb-5"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="btn btn-primary pointer btn-block"
        type="submit"
      >
        Login
      </button>
      <hr />
      <div
        class="auth__social-networks"
      >
        <p>
          Login With Social Networks
        </p>
        <div
          class="google-btn"
        >
          <div
            class="google-icon-wrapper"
          >
            <img
              alt="google button"
              class="google-icon"
              src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
            />
          </div>
          <p
            class="btn-text"
          >
            <b>
              Sign in with google
            </b>
          </p>
        </div>
      </div>
      <a
        class="link"
        href="/auth/register"
      >
        Create New Account
      </a>
    </form>
  </div>
</div>
`;
